### DISCUSSION TITLE
Some questions about USB device and HID

### DISCUSSION THREAD

**Discussion by occivink** [hathach/tinyusb#1484] [Q&A]
Hi, I've got some questions about using tinyUSB for a HID device. For context I'm building a keyboard where I'm using raspberry pi picos as the microcontrollers (and therefore tinyUSB). The current firmware can be found [here](https://github.com/occivink/Keyboard/blob/master/firmware/src/main.c). For what it's worth the code is already functional (in that it behaves like a keyboard with the programmed layout) but I would like some clarifications regarding the USB communication.

As part of the examples, this comment can often be read regarding `tud_suspend_cb(bool remote_wakeup_en)`:
>Within 7ms, device must draw an average of current less than 2.5 mA from bus

However, none of the examples seem to do anything meaningful in this callback. 
* How is one supposed to reduce the power draw? I'm not even sure how much power I'm currently drawing, and how realistic of a target this is.
* Are there ways to measure this that do not require expensive equipment?
* What might happen if the device still draws more than 2.5mA?
* What determines whether remote wakeup is enabled?
* What can trigger a suspend callback? Is this when the host is suspended, or can it happen at any point (for example, if the host thinks the device is not doing anything)?

I realize that some of these questions are maybe more relevant to parts of the stack that are outside of tinyUSB, but I thought you might know.

All HID examples use `tud_hid_n_report()` (or some wrapper around it), and in all of them the `tud_hid_get_report_cb()` is unimplemented and simply returns 0 (with a TODO comment).
* As far as I understand, these mechanisms serve the same purpose, it's just that one is synchronous while the other happens in a callback. Is that correct? 
* Is there a reason to prefer one over the other?
* Is `tud_hid_n_report()` actually synchronously sending the report, or does it schedule it to be sent on the next `tud_task()`?
* The callback interface states that 'returning 0 will cause the stack to STALL request'. What does stalling mean? Is it OK to stall indefinitely?
* Is it better, worse or equivalent to always send reports at regular intervals (even if they're the same) vs. only sending them if their content is different? (aka always vs. on keypress/release)

**Comment by hathach** [comment]
> * How is one supposed to reduce the power draw? I'm not even sure how much power I'm currently drawing, and how realistic of a target this is.
> * Are there ways to measure this that do not require expensive equipment?
> * What might happen if the device still draws more than 2.5mA?
> * What determines whether remote wakeup is enabled?
> * What can trigger a suspend callback? Is this when the host is suspended, or can it happen at any point (for example, if the host thinks the device is not doing anything)?

There is lots of cheap current measurement board such as nordic ppk2 that could measure uA. When pc/laptop/ipad enter suspend mode, and your device violate the power, depending on the host driver, it may cut off the power to usb device. Callback is triggered when usb bus in suspended mode i.e pc is in sleep.

> All HID examples use `tud_hid_n_report()` (or some wrapper around it), and in all of them the `tud_hid_get_report_cb()` is unimplemented and simply returns 0 (with a TODO comment).
> 
> * As far as I understand, these mechanisms serve the same purpose, it's just that one is synchronous while the other happens in a callback. Is that correct?
> * Is there a reason to prefer one over the other?
> * Is `tud_hid_n_report()` actually synchronously sending the report, or does it schedule it to be sent on the next `tud_task()`?
> * The callback interface states that 'returning 0 will cause the stack to STALL request'. What does stalling mean? Is it OK to stall indefinitely?
> * Is it better, worse or equivalent to always send reports at regular intervals (even if they're the same) vs. only sending them if their content is different? (aka always vs. on keypress/release)

these are generic questions, please check the source code and/or usb specs for the detail
